cfscrape/__init__.py

```diff
@@ -174,60 +187,66 @@ def solve_cf_challenge(self, resp, **original_kwargs):
 
         # Cloudflare requires a delay before solving the challenge
         time.sleep(max(delay - (time.time() - start_time), 0))
 
         # Send the challenge response and handle the redirect manually
         redirect = self.request(method, submit_url, **cloudflare_kwargs)
-        redirect_location = urlparse(redirect.headers["Location"])
-
-        if not redirect_location.netloc:
-            redirect_url = urlunparse(
-                (
-                    parsed_url.scheme,
-                    domain,
-                    redirect_location.path,
-                    redirect_location.params,
-                    redirect_location.query,
-                    redirect_location.fragment,
+        if "Location" in redirect.headers:
+            redirect_location = urlparse(redirect.headers["Location"])
+
+            if not redirect_location.netloc:
+                redirect_url = urlunparse(
+                    (
+                        parsed_url.scheme,
+                        domain,
+                        redirect_location.path,
+                        redirect_location.params,
+                        redirect_location.query,
+                        redirect_location.fragment,
+                    )
                 )
-            )
-            return self.request(method, redirect_url, **original_kwargs)
-        return self.request(method, redirect.headers["Location"], **original_kwargs)
+                return self.request(method, redirect_url, **original_kwargs)
+            return self.request(method, redirect.headers["Location"], **original_kwargs)
+        else:
+            return self.request(self.org_method, submit_url, **cloudflare_kwargs)
+
 
     def solve_challenge(self, body, domain):
         try:
+            tree_ = etree.fromstring(body, etree.HTMLParser())
             challenge, ms = re.search(
                 r"setTimeout\(function\(\){\s*(var "
                 r"s,t,o,p,b,r,e,a,k,i,n,g,f.+?\r?\n[\s\S]+?a\.value\s*=.+?)\r?\n"
                 r"(?:[^{<>]*},\s*(\d{4,}))?",
-                body,
+                tree_.xpath('//script[@type="text/javascript"]/text()')[0],
             ).groups()
 
             # The challenge requires `document.getElementById` to get this content.
             # Future proofing would require escaping newlines and double quotes
-            innerHTML = re.search(r"<div(?: [^<>]*)? id=\"cf-dn.*?\">([^<>]*)", body)
-            innerHTML = innerHTML.group(1) if innerHTML else ""
+            innerHTML = ''
+            for i in tree_.xpath('//script[@type="text/javascript"]/text()')[0].split(';'):
+                if i.strip().split('=')[0].strip() == 'k':      # from what i found out from pld example K var in
+                    k = i.strip().split('=')[1].strip(' \'')    #  javafunction is for innerHTML this code to find it
+                    innerHTML = tree_.xpath('//form[@id="challenge-form"]/../div[@id="' + k + '"]/text()')[0]
```

Comment:
by the way i replaced most re function with lxml.etree
because even tough re is good for matching  pattern its still ridged in html while lxml.etree move relatively in html 
like how in line 244 i looked for "challenge-form" then moved up to go to next div that has innerHTML